async-book/src/execution/future.md

```diff
@@ -0,0 +1,178 @@
+            //
+            // Arrange for `wake` to be called once data is available.
+            // When data becomes available, `wake` will be called, and the
+            // user of this `Future` will know to call `poll` again and
+            // receive data.
+            self.socket.set_readable_callback(wake);
+        }
+    }
+}
+```
+
+This model of `Future`s allows for composing together multiple asynchronous
+operations without needing intermediate allocations. Running multiple futures
+at once or chaining futures together can be implemented via allocation-free
+state machines, like this:
+
+```rust
+/// A SimpleFuture that runs two other futures to completion concurrently.
+///
+/// Concurrency is acheived via the fact that calls to `poll` each future
+/// may be interleaved, allowing each future to advance itself at its own pace.
+struct Join2 {
+    // Each field may contain a future that should be run to completion.
+    // If the future has already completed, the field is set to `None`.
+    a: Option<FutureA>,
+    b: Option<FutureB>,
+}
+
+impl SimpleFuture for Join2 {
+    type Output = ();
+    fn poll(&mut self, wake: fn()) -> Poll<Self::Output> {
+        // Attempt to complete future `a`.
+        let finished_a = match &mut self.a {
+            Some(a) => {
+                match a.poll(wake) {
+                    Poll::Ready(()) => true,
+                    Poll::Pending => false,
+                }
+            }
+            None => true,
+        };
+        if finished_a { self.a.take() }
+
+        // Attempt to complete future `b`.
+        let finished_b = match &mut self.b {
+            Some(b) => {
+                match b.poll(wake) {
+                    Poll::Ready(()) => true,
+                    Poll::Pending => false,
+                }
+            }
+            None => true,
+        };
+        if finished_b { self.b.take() }
+
+        if finished_a && finished_b {
+            // Both futures have completed-- we can return successfully
+            Poll::Ready(())
+        } else {
+            // One or both futures still have work to do, and will call
+            // `wake()` when progress can be made.
+            Poll::Pending
+        }
+    }
+}
+```
+
+This shows how multiple futures can be run simultaneously without needing
+separately allocated tasks, allowing for more efficient asynchronous programs.
+Similarly, multiple sequential futures can be run one after another, like this:
+
+```rust
+/// A SimpleFuture that runs two futures to completion, one after another.
+//
+// Note: for the purposes of this simple example, `AndThenFut` assumes both
+// the first and second futures are available at creation-time. The real
+// `AndThen` combinator allows creating the second future based on the output
+// of the first future, like `get_breakfast.and_then(|food| eat(food))`.
+enum AndThenFut {
+    first: Option<FutureA>,
+    second: FutureB,
+}
+
+impl SimpleFuture for Join2 {
```

Comment:
Shouldn't `SimpleFuture` be implemented for `AndThenFut`?

Reply:
Fixed the name typo.

RE the second concern: if the first future returns pending, the second future won't be polled because we'll hit the `Poll::Pending => return Poll::Pending` branch. Does that make sense, or do you think some other explanation is necessary?

Reply:
Oh yeah, I completely missed the early return. My bad!